torch/csrc/DeviceAccelerator.cpp

```diff
@@ -0,0 +1,79 @@
+#include <c10/core/DeviceGuard.h>
+#include <torch/csrc/DeviceAccelerator.h>
+#include <torch/csrc/utils/device_lazy_init.h>
+
+namespace torch::accelerator {
+
+void initModule(PyObject* module) {
+  auto m = py::handle(module).cast<py::module>();
+
+  m.def("_accelerator_getAccelerator", []() {
+    // If no accelerator is currently available, return CPU.
+    return c10::Device(at::getAccelerator(false).value_or(c10::kCPU));
+  });
+
+  m.def("_accelerator_deviceCount", []() {
+    const auto device_type = at::getAccelerator(false);
+    if (!device_type.has_value()) {
+      return static_cast<c10::DeviceIndex>(0);
+    }
+    torch::utils::maybe_initialize_device(device_type.value());
```

Comment:
I don't recall the device initialization logic off the top of my head but why do we need to do this here and it is not done in the impl below?
What is the rule for which layer in our stack is responsible for calling the device init?

Reply:
Let's take cuda as an example. [`maybe_initialize_device`](https://github.com/pytorch/pytorch/blob/cdc9f14227f68a3633ae1af44632242fe8e8b9aa/torch/csrc/utils/device_lazy_init.h#L29) will invoke [`torch.cuda._lazy_init`](https://github.com/pytorch/pytorch/blob/cdc9f14227f68a3633ae1af44632242fe8e8b9aa/torch/cuda/__init__.py#L289). When user type `import torch`, cuda underlying runtime should not be called at this moment. Otherwise, the process later forked will be poisoned. I believe this issue is not CUDA-specific, each backend will meet this limitation. Up to now, there are 5 backends supported [`maybe_initialize_device`](https://github.com/pytorch/pytorch/blob/main/torch/csrc/utils/device_lazy_init.h#L31-L32)
So PyTorch uses lazy initialization mechanism to guarantee the usage. In other worlds, PyTorch only invokes the cuda runtime API in the first place where it is needed.
> What does `torch.cuda._lazy_init` do?

1. initialize some instances per device, like [`CUDACachingAllocator`](https://github.com/pytorch/pytorch/blob/cdc9f14227f68a3633ae1af44632242fe8e8b9aa/c10/cuda/CUDACachingAllocator.cpp#L3222) is invoked [here](https://github.com/pytorch/pytorch/blob/cdc9f14227f68a3633ae1af44632242fe8e8b9aa/torch/csrc/cuda/Module.cpp#L1428),
2. initialize the process poisoned flag [here](https://github.com/pytorch/pytorch/blob/cdc9f14227f68a3633ae1af44632242fe8e8b9aa/torch/csrc/cuda/Module.cpp#L1427), and check if the forked process is poisoned [here](https://github.com/pytorch/pytorch/blob/cdc9f14227f68a3633ae1af44632242fe8e8b9aa/torch/cuda/__init__.py#L304). Raising a user-friendly message if fork process is poisoned.

> Which layer in our stack is responsible for calling the device init?

The main rule is the lazy initialization must be called before any API in PyTorch that requires cuda underlying runtime. For example:
1. some Python API, like [`torch.cuda.set_device`](https://github.com/pytorch/pytorch/blob/c0879d0c2111ffa9efed66225b4fe815c4641d47/torch/csrc/cuda/Module.cpp#L90).
2. new tensor created, like [here](https://github.com/pytorch/pytorch/blob/cdc9f14227f68a3633ae1af44632242fe8e8b9aa/torch/csrc/utils/tensor_new.cpp#L74)
3. some factory function codegen, like [here](https://github.com/pytorch/pytorch/blob/cdc9f14227f68a3633ae1af44632242fe8e8b9aa/tools/autograd/templates/python_variable_methods.cpp#L1001)
4. some ops codegen, like [here](https://github.com/pytorch/pytorch/blob/cdc9f14227f68a3633ae1af44632242fe8e8b9aa/torchgen/api/python.py#L1484)

In summary, I believe this is a must and needs to be placed out of C++ impl because it is related to Python.